scripts/surface_charge/surface_charge_calculator.py

```diff
@@ -29,30 +28,107 @@ def __init__(self, crystal, use_existing_charges: bool = False):
+        triangle_areas = self.calculate_area_of_triangles(list(self.surface.topology.triangles))
+        total_triangle_area = sum(triangle_areas)
+        for node_index, triangle_area in zip(tri_index, triangle_areas):
+            average_triangle_charge = np.mean([self.node_charge_dictionary[i] for i in node_index])
+            triangle_representation = triangle_area / surface_area
+            projected_charge = 0
+            if np.isclose(triangle_area, 0.0) is False:
+                projected_charge = average_triangle_charge / triangle_area
+            self.triangles_properties[tuple(node_index)] = {'Average Charge': average_triangle_charge,
+                                                            'Triangle Area': triangle_area,
+                                                            'Percentage Area': triangle_representation,
+                                                            'Node Representative Charge': average_triangle_charge * triangle_representation,
+                                                            'Node Projected Surface Charge': projected_charge}
+
+    def calculate_node_charges(self):
+        tri_index = self.calculated_node_index_values(list(self.surface.topology.nodes),
+                                                      list(self.surface.topology.triangles))
+        self.get_average_node_charge()
+        self.calculate_triangles_properties(tri_index)
+        self.representative_charge = np.sum(
+            [triangle['Node Representative Charge'] for triangle in self.triangles_properties.values()])
+        self.node_charges = np.sum([triangle['Average Charge'] for triangle in self.triangles_properties.values()])
+        return self.representative_charge
+
+    @staticmethod
+    def calculate_length(origin: np.ndarray, target: np.ndarray) -> float:
+        """Returns distance between target and origin"""
+        if not isinstance(origin, np.ndarray) or not isinstance(target, np.ndarray):
+            raise TypeError("Please supply numpy arrays for lengths.")
+        return np.linalg.norm(target - origin)
+
+    @staticmethod
+    def compute_triangle_area(a: float, b: float, c: float) -> float:
+        """Calculates area of triangle using Heron's formula"""
+        s = (a + b + c) / 2
+        return np.sqrt(s * (s - a) * (s - b) * (s - c))
+
+    def calculate_area_of_triangles(self, triangles: List) -> List:
+        """ Calculates area of individual triangles from node positions using Heron's formula"""
+        triangle_areas = []
+        for triangle in triangles:
+            pos_0, pos_1, pos_2 = np.array(triangle[0]), np.array(triangle[1]), np.array(triangle[2]),
+            a_dist = self.calculate_length(pos_0, pos_1)
+            b_dist = self.calculate_length(pos_0, pos_2)
+            c_dist = self.calculate_length(pos_1, pos_2)
+            triangle_areas.append(self.compute_triangle_area(a_dist, b_dist, c_dist))
+
+        return triangle_areas
```

Comment:
`
# the formula for determining the area of a triangle of three points in 3 dimensions
def area_tri(a, b, c):
    return np.linalg.norm(np.cross((b - a), (c - a))) / 2


# the area of a polygon using the above formula
def area_poly(arr):
    a = arr[1:-1] - arr[0]
    b = arr[2:] - arr[0]
    return np.sum(np.linalg.norm(np.cross(a, b, axis=1), axis=1)) / 2`

much more elegant

Reply:
this is from the voronoi code, btw